### PR TITLE
Add boolean finder

### LINKED ISSUE #18: Boolean Analysis
Build a CFG analysis pass that builds a set of `VariableName` which are known to be booleans (the result of any logical comparison).

This will need to run after SSA has been applied.

### PR DESCRIPTION
**Pull request by thomasio101**
This pull request resolves #18 by adding adding the means to find all boolean variables in a Yolol program.

### CODE REVIEW

**Comment by martindevans** [comment]
There is another case that can be considered here. Consider this code:

```
a = :b / 2 goto 3
a = "string" goto 3
:c = a goto 3
```

If `:b` is a number then we skip straight to line 3, if `:b` was a string we execute line 2 first.

Rewritten into SSA form it will look like this:

```
a[0] = :b / 2 goto 3
a[1] = "string" goto 3
:c = Phi(a[0], a[1]) goto 3
```

The `Phi` node means it is reading _either_ `a[0]` or `a[1]` depending upon the execution path taken to get here.

Your boolean analysis could handle `Phi` nodes - if _all_ of the variables which `Phi` accesses are marked as booleans, you can consider the `Phi` node to be a boolean.

File: YololEmulator.Tests/Analysis/Inspection/FindBooleans.cs [hunk 1]
```diff
@@ -31,10 +31,52 @@ public void FindBooleansSingleLine()
...
+                "a=1 theAnswer=42 b=theAnswer==a"
+            );
+
+            var cfg = new Yolol.Analysis.ControlFlowGraph.Builder(ast.StripTypes()).Build();
+
+            cfg = cfg.StaticSingleAssignment(out var ssa);
+
+            var variableNames = cfg.FindBooleanVariables(ssa).ToHashSet();
+
+            Assert.IsTrue(variableNames.SetEquals(new VariableName[] {
```
Comment (martindevans) [inline YololEmulator.Tests/Analysis/Inspection/FindBooleans.cs]: This test relies on the naming scheme of the AST builder, which is an unstable implementation detail. It's not critical to fix that in this PR.

Comment (thomasio101) [reply YololEmulator.Tests/Analysis/Inspection/FindBooleans.cs]: I'm taking a look at it now. ~I feel like I might have a solution.~ I have already found the solution and am currently cleaning it up.

Comment (martindevans) [reply YololEmulator.Tests/Analysis/Inspection/FindBooleans.cs]: Btw the naming of `a[0]` and `b[0]` ought to be stable, it's just the intermediate values that may change (that's dependent upon whatever the ExpressionDecomposition pass does). So you could just validate that `a[0]` and `b[0]` get classified as booleans.

File: Yolol.Analysis/ControlFlowGraph/Extensions/AnalysisExtensions.cs [hunk 2]
```diff
@@ -70,5 +71,38 @@ public static class AnalysisExtensions
...
+                cfg.VisitBlocks(() => new FindBooleanVariables(booleans, nonBooleans, unknowns, ssa));
+            }
+
+            var result = new HashSet<VariableName>(booleans.Where(n => !n.IsExternal));
+
+            return result;
+        }
+
+        [NotNull]
+        public static IReadOnlyCollection<Assignment> FindVariableAssignments([NotNull] this IControlFlowGraph cfg, ISingleStaticAssignmentTable ssa)
```
Comment (martindevans) [inline Yolol.Analysis/ControlFlowGraph/Extensions/AnalysisExtensions.cs]: It is not clear from the name of this method that it does not find external assignments. I can see three ways to resolve this:
 - Rename to `FindNonExternalVariableAssignments`
 - Make it `private`
 - Add a `bool includeExternals = false` parameter

My preference is `private`

File: Yolol.Analysis/TreeVisitor/Inspection/FindBooleanVariables.cs [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+        public FindBooleanVariables(HashSet<VariableName> booleanNames, HashSet<VariableName> nonBooleanNames, HashSet<VariableName> unknownNames, ISingleStaticAssignmentTable ssa)
+        {
+            _booleanNames = booleanNames;
+            _nonBooleanNames = nonBooleanNames;
+            _unknownNames = unknownNames;
+            _ssa = ssa;
+        }
+
+        protected override BaseStatement Visit(Assignment ass)
```
Comment (martindevans) [inline Yolol.Analysis/TreeVisitor/Inspection/FindBooleanVariables.cs:34]: `Phi` expressions not handled, as discussed in general PR comments.

File: YololEmulator.Tests/Analysis/Inspection/FindBooleans.cs [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+using Yolol.Grammar.AST;
+using Yolol.Grammar.AST.Expressions;
+using Yolol.Grammar.AST.Statements;
+
+namespace YololEmulator.Tests.Analysis.Inspection
+{
+    [TestClass]
+    public class FindBooleans
+    {
+        private void _test(Program ast, IEnumerable<VariableName> originalBooleanVariableNames)
```
Comment (martindevans) [inline YololEmulator.Tests/Analysis/Inspection/FindBooleans.cs]: private methods should be named `Test`, only private fields are named with leading underscores.

**Review by martindevans — CHANGES_REQUESTED** [review]

File: Yolol.Analysis/ControlFlowGraph/Extensions/AnalysisExtensions.cs (hunk 2 above)
Comment (thomasio101) [reply Yolol.Analysis/ControlFlowGraph/Extensions/AnalysisExtensions.cs]: I'm going for `includeExternals = false` as this is simple and encourages code reuse.

File: Yolol.Analysis/TreeVisitor/Inspection/FindBooleanVariables.cs (hunk 3 above)
Comment (thomasio101) [reply Yolol.Analysis/TreeVisitor/Inspection/FindBooleanVariables.cs:34]: Getting right on it!

File: YololEmulator.Tests/Analysis/Inspection/FindBooleans.cs (hunk 4 above)
Comment (thomasio101) [reply YololEmulator.Tests/Analysis/Inspection/FindBooleans.cs]: I hadn't noticed that pattern, but I'll change that now.

**Comment by martindevans** [comment]
Since this seems dead I've merged the simpler alternative I proposed (#22). I'm still interested in any tests you have which prove this approach is superior :)

**Pull request closed without merging**